### PR TITLE
Manage the pipeline lifecycle in a GenServer

### PR DESCRIPTION
**Pull request by garrettmichaelgeorge**
This provides the first concurrent implementation of a Pipeline Runner, responsible for managing the lifecycle of a single pipeline, modeled as a GenServer that runs each step in sequence and terminates gracefully.

## Tracking step progress with a zipper

Previously, a pipeline's steps were modeled as a list, but that did not allow for tracking which steps had been run and which one is currently running. Neither does a queue.

After some reading, I think a better model for tracking the progression of steps is that of a [zipper](https://en.wikipedia.org/wiki/Zipper_(data_structure)) that can keep track of where the "focus" is at any given time. I implemented it as a simple `ExIntegrate.Core.Zipper` data structure in two dimensions. 

So calling `my_pipeline.steps` returns a zipper that contains all the steps that have run, the step currently in focus, and the steps remaining. Public functions on `ExIntegrate.Core.Pipeline` wrap the bare Zipper operations and expose all the operations needed for managing pipeline state. For instance, advance the pipeline forward to the next step with `Pipeline.advance/1`, or check if all steps have completed with `Pipeline.completed?/1`. Under the hood, it relies on the zipper implementation.

References:
  * Wikipedia [Zipper](https://en.wikipedia.org/wiki/Zipper_(data_structure)) article
  * [Zipper article](https://github.com/doorgan/sourceror/blob/main/notebooks/zippers.livemd) by Doorgan as well as the [source code](https://github.com/doorgan/sourceror/blob/main/lib/sourceror/zipper.ex) for his own zipper implementation (a three-dimensional one for traversing the Elixir AST).
  * [Gérard Huet's original paper](https://www.st.cs.uni-saarland.de/edu/seminare/2005/advanced-fp/docs/huet-zipper.pdf)
  * Zipper module from the [Clojure stdlib](https://clojuredocs.org/clojure.zip/zipper)
  * [ElixirForum discussion](https://elixirforum.com/t/elixir-needs-a-fifo-type/5701/24)

## Lifecycle of a PipelineRunner
1. Spawned using `PipelineRunner.start_link/1`, with a `%Pipeline{}` passed in for the initial state
2. Automatically starts the first step, wrapping it in a call to `ExIntegrate.TaskSupervisor.async_nolink/3`
3. Listen for a message with the step results
  a. On step success, stores the step results and starts the next step.
  b. On step failure, terminates without starting any more steps
    - We may want to customize this behavior
4. On successful completion of all steps, terminates with `:normal` [exit status](https://hexdocs.pm/elixir/Supervisor.html#module-exit-reasons-and-restarts)

Note that when PipelineRunner terminates for any reason (i.e. 3b or 4), it will *not* be restarted by its parent supervisor, owing to its child spec [restart value](https://hexdocs.pm/elixir/Supervisor.html#module-restart-values-restart) of `:temporary`. I believe this is the desired behavior, but please correct me if I am missing something.

## Final thoughts
I haven't covered every edge case here. We'll want to consider at least the following scenarios concerning individual pipeline runs:
- A particular step takes too long and times out
- The PipelineRunner is shut down mid-step by an external process (like a supervisor), or even by application termination or other unforeseen causes

WDYT? @samrose

### CODE REVIEW

**Comment by garrettmichaelgeorge** [comment]
FWIW I used `Logger` throughout the different phases of running a pipeline, but mostly as a placeholder until we look more closely at how to handle logging.

File: lib/ex_integrate.ex [hunk 1]
```diff
@@ -4,11 +4,12 @@ defmodule ExIntegrate do
   """
 
   alias ExIntegrate.Boundary.PipelineRunner
```
Comment (samrose) [inline lib/ex_integrate.ex:6]: @garrettmichaelgeorge  why "boundary"? Just curious

File: lib/ex_integrate/pipeline_runner.ex [hunk 2]
```diff
@@ -2,14 +2,82 @@ defmodule ExIntegrate.Boundary.PipelineRunner do
...
+    {:stop, :step_failure, state}
+  end
+
+  @impl GenServer
+  def handle_info({:DOWN, ref, _, _, reason}, {state, config}) do
+    Logger.info("Step task #{inspect(ref)} terminated unexpectedly. Reason: #{inspect(reason)}")
+    {:stop, :step_failure, state}
+  end
+
+  @deprecated """
```
Comment (samrose) [inline lib/ex_integrate/pipeline_runner.ex:75]: It seems like this could just be dropped so that we don't carry forward dead code.

File: lib/ex_integrate/zipper.ex [hunk 3]
```diff
@@ -0,0 +1,82 @@
+defmodule ExIntegrate.Core.Zipper do
```
Comment (samrose) [inline lib/ex_integrate/zipper.ex:1]: interesting solution RE: zipper. I guess this lets you get more arbitrary than the tools libgraph offers to move around the graph eh?

this makes sense as I see in this context that Libgraph doesn't really have a way to specifically query the graph. It does have algorithms for searching, or algorithmic processing, but nothing that would let you pinpoint positions on it

**Review by samrose — APPROVED** [review]

**Comment by samrose** [comment]
It all looks really good. I am going to test locally just for fun, and then merge it

**Comment by samrose** [comment]
@garrettmichaelgeorge wrt to

> I haven't covered every edge case here. We'll want to consider at least the following scenarios concerning individual pipeline runs:
> 
>     A particular step takes too long and times out
>     The PipelineRunner is shut down mid-step by an external process (like a supervisor), or even by application termination or other unforeseen causes
> 
> 

I think we should consider giving the author of an ei.sjon file some limited but predictable ways to respond to these outcomes via a k:v pair per step. But I think we should discuss and explore before just diving in and coding this, so that we can think through the simplest rules that can cover this and integrate well with the OTP approach and also cover the cases most CI/CD users would run into + we could have a default that makes the most sense as well.

**Comment by samrose** [comment]
@garrettmichaelgeorge  in the case of 

> The PipelineRunner is shut down mid-step by an external process (like a supervisor), or even by application termination or other unforeseen causes

pretty much all CI/CD systems would fail this build if they can, and insert the reason being that the CI/CD system had an fatal error (considering this fatal to the intended outcome because it was unexpected and broke functionality)

**Comment by samrose** [comment]
@garrettmichaelgeorge in the case of timeouts, the best thing to do is create a default that users can know and expect, and in each step offer an option override

**Comment by samrose** [comment]
One feature that we still need to account for is to respond to the stdout, stderr status returned by Rambo, so that if one pipeline depends on another, but the first has failed with status 1, then the run of pipelines should stop

example

```Rambo.run("mix", "compile",log: &IO.inspect/1)
{:ok, %Rambo{err: "", out: "", status: 0}}
iex(5)> Rambo.run("mix", "compil0",log: &IO.inspect/1)
{:stderr, "** (Mix) The task \"compil0\" could not be found. Did you mean \"co"}
{:stderr, "mpile\"?\n"}
{:error,
 %Rambo{
   err: "** (Mix) The task \"compil0\" could not be found. Did you mean \"compile\"?\n",
   out: "",
   status: 1
 }}
```

Right now in a test ei.json if the 2nd commend is put into the ei.json as a step in a pipeline, and the command fails with status 1, then ex_integrate still proceeds with the rest of the run

File: lib/ex_integrate.ex (hunk 1 above)
Comment (garrettmichaelgeorge) [reply lib/ex_integrate.ex:6]: @samrose we had discussed adopting the conventions from *[Designing Elixir Systems with OTP](https://pragprog.com/titles/jgotp/designing-elixir-systems-with-otp/)*, where they namespace their functional core modules under `MyApp.Core.*` and their boundary modules under `MyApp.Boundary.* Since the PipelineRunner ends up performing IO operations by running steps, I treated it as a boundary module (a Pipeline itself, by contrast, is under `ExIntegrate.Core.Pipeline`.

But I have been a bit ambivalent about that namespacing idea lately. Outside that book, I haven't seen a single Elixir application that uses this convention. And while "core" and "boundary" are useful terms for considering the layers of the application, I don't know that they necessarily need to be enshrined in the code.

On the other hand, the conceptual separation between core and boundary modules has helped keep the code clean so far.

I had even considered using `ExIntegrate.Core.*` and `ExIntegrate.Boundary.*` as scaffolding until we got the main app structure the way we wanted, then removing the namespaces and simply using `ExIntegrate.Pipeline` and `ExIntegrate.PipelineRunner`.

If we were to follow the original idea through, we would probably want to move the files into `lib/ex_integrate/core/` and `lib/ex_integrate/boundary/` to match up with the module names.

Happy to discuss it more…I can't quite make up my mind about it.

File: lib/ex_integrate/pipeline_runner.ex (hunk 2 above)
Comment (garrettmichaelgeorge) [reply lib/ex_integrate/pipeline_runner.ex:75]: Agreed in principle. I guess my reason for deprecating was that the top-level `ExIntegrate`, and its corresponding tests, still utilize this sequential function to run a series of pipelines. If we removed this code now, we wouldn't be able to run our `ei.json` file anymore until we finished the rest of the OTP structure.

Stepping back, I guess we initially scaffolded a minimal, sequential implementation that could complete an entire run from an `ei.json` file, and now we are partway through the process of replacing that implementation with a concurrent one.

What remains is adding the higher-level modules: a RunManager for launching and responding to PipelineRunners, and a long-lived `ExIntegrate.Server` process that starts runs on demand by launching a RunManager.

If I understand correctly, once the OTP structure is in place, we can safely remove this code.

File: lib/ex_integrate.ex (hunk 1 above)
Comment (samrose) [reply lib/ex_integrate.ex:6]: @garrettmichaelgeorge no that is great, I just spaced out on the origin of it so thanks

File: lib/ex_integrate/pipeline_runner.ex (hunk 2 above)
Comment (samrose) [reply lib/ex_integrate/pipeline_runner.ex:75]: @garrettmichaelgeorge makes sense now thanks!
